controllers/lease_controller.go

```diff
@@ -0,0 +1,544 @@
+		full, err := strconv.ParseInt(*fullLease.Spec.HolderIdentity, 10, 64)
+		if err != nil {
+			logger.Error(err, "Can't convert holder identity of full snap lease to integer")
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, err
+		}
+
+		delta, err := strconv.ParseInt(*deltaLease.Spec.HolderIdentity, 10, 64)
+		if err != nil {
+			logger.Error(err, "Can't convert holder identity of delta snap lease to integer")
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, err
+		}
+
+		diff := delta - full
+
+		// Reconcile job only when current revision is more than 1 million (this vale is configurable through `events-threshold` option in druid) than
+		// the last snapshot (full/compact) revision
+		if diff >= lc.config.EventsThreshold {
+			return lc.reconcileJob(ctx, logger, etcd)
+		}
+	}
+
+	return ctrl.Result{
+		Requeue: false,
+	}, nil
+}
+
+func (lc *LeaseController) reconcileJob(ctx context.Context, logger logr.Logger, etcd *druidv1alpha1.Etcd) (ctrl.Result, error) {
+	logger.Info("Reconcile etcd compaction job")
+
+	// First check if a job is already running
+	job := &batchv1.Job{}
+	err := lc.Get(ctx, types.NamespacedName{Name: getJobName(etcd), Namespace: etcd.Namespace}, job)
+
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, fmt.Errorf("error while fetching compaction job: %v", err)
+		}
+		// Required job doesn't exist. Create new
+		job, err = lc.createCompactJob(ctx, logger, etcd)
+		logger.Info("Job Creation")
+		if err != nil {
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, fmt.Errorf("error during compaction job creation: %v", err)
+		}
+	}
+
+	logger.Info("Current compaction job is: " + job.Name + ", status: " + strconv.FormatInt((int64(job.Status.Succeeded)), 32))
+
+	// Delete job and requeue if the job failed
+	if job.Status.Failed > 0 {
+		err = lc.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil {
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, fmt.Errorf("error while deleting failed compaction job: %v", err)
+		}
+		return ctrl.Result{
+			RequeueAfter: 10 * time.Second,
+		}, fmt.Errorf("existing job status failed")
+	}
+
+	// Delete job and return if the job succeeded
+	if job.Status.Succeeded > 0 {
+		return lc.delete(ctx, logger, etcd)
+	}
+
+	return ctrl.Result{Requeue: false}, nil
+}
+
+func (lc *LeaseController) delete(ctx context.Context, logger logr.Logger, etcd *druidv1alpha1.Etcd) (ctrl.Result, error) {
+	job := &batchv1.Job{}
+	err := lc.Get(ctx, types.NamespacedName{Name: getJobName(etcd), Namespace: etcd.Namespace}, job)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, fmt.Errorf("error while fetching compaction job: %v", err)
+		}
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	if job.Name == getJobName(etcd) && job.DeletionTimestamp == nil {
```

Comment:
Right.  Problem is, `DeletionTimestamp` can be nil both for job which was found and job which was not found (i.e empty jobs. See the snippet before this call). So, `job.Name == getJobName(etcd)` seemed to be a good trick to find if the job was found. Otherwise, even if a job was not found, then also it was trying to enter into the `if` block and throwing error

Reply:
Hm, I don't understand. When the job was not found, then the function is already left before in line [225](https://github.com/gardener/etcd-druid/pull/235/files/8f3d21beb570ef785d44340e6d55e5aceeb985aa#diff-a51e70a68174a40a7e3aa8b6bc4445c4327005a6734c13d07e0f615ed40765c6R225), right?

Reply:
Sorry! My bad. I misunderstood something

Reply:
Removed